Nightly clang/7+cuda/9.2 compilation errors with -Werror: Kokkos_Array.hpp:354:1: error: 'tuple_size' defined as a struct template here but previously declared as a class template

Following merge of #4962 the clang/7+cuda/9.2 nightly job showed compilation errors:

```
22:23:31 In file included from /home/jenkins/jenkins-new/workspace/Kokkos_develop_kokkos-dev_clang_701_cuda_92/kokkos/core/src/impl/Kokkos_ViewArray.hpp:48:
22:23:31 /home/jenkins/jenkins-new/workspace/Kokkos_develop_kokkos-dev_clang_701_cuda_92/kokkos/core/src/Kokkos_Array.hpp:354:1: error: 'tuple_size' defined as a struct template here but previously declared as a class template [-Werror,-Wmismatched-tags]
22:23:31 struct std::tuple_size<Kokkos::Array<T, N>>
22:23:31 ^
22:23:31 /projects/sems/install/rhel7-x86_64/sems/compiler/gcc/7.2.0/base/lib/gcc/x86_64-pc-linux-gnu/7.2.0/../../../../include/c++/7.2.0/array:342:5: note: did you mean struct here?
22:23:31     class tuple_size;
22:23:31     ^
22:23:31 In file i
```

**Reproducer**:
```
module load sems-archive-env sems-archive-cmake/3.17.1 sems-archive-gcc/7.2.0 sems-archive-cuda/9.2 sems-archive-clang/7.0.1

$KOKKOS_PATH/generate_makefile.bash --with-devices=Cuda,OpenMP --arch=Kepler35 --compiler=clang++ --cxxflags="-O3 -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized " --cxxstandard="14" --kokkos-path=$KOKKOS_PATH --no-examples
```

--- a/core/src/Kokkos_Array.hpp
+++ b/core/src/Kokkos_Array.hpp
@@ -350,6 +350,10 @@ struct Array<T, KOKKOS_INVALID_INDEX, Array<>::strided> {
 // guarding against bogus error 'specialization in different namespace' with
 // older GCC that do not support C++17 anyway
 #if !defined(KOKKOS_COMPILER_GNU) || (KOKKOS_COMPILER_GNU >= 710)
+#if defined(KOKKOS_COMPILER_CLANG) && KOKKOS_COMPILER_CLANG < 800
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wmismatched-tags"
+#endif
 template <class T, std::size_t N>
 struct std::tuple_size<Kokkos::Array<T, N>>
     : std::integral_constant<std::size_t, N> {};
@@ -358,6 +362,9 @@ template <std::size_t I, class T, std::size_t N>
 struct std::tuple_element<I, Kokkos::Array<T, N>> {
   using type = T;
 };
+#if defined(KOKKOS_COMPILER_CLANG) && KOKKOS_COMPILER_CLANG < 800
+#pragma clang diagnostic pop
+#endif
 #endif
 
 namespace Kokkos {

